internal/cluster/connection_manager.go

```diff
@@ -520,89 +524,106 @@ func (m *ConnectionManager) detectFixBrokenConnections() {
 }
 
 type connectionMap struct {
-	connectionsMu *sync.RWMutex
+	lb pubcluster.LoadBalancer
+	mu *sync.RWMutex
 	// addrToConn maps connection address to connection
 	addrToConn map[pubcluster.Address]*Connection
 	// connToAddr maps connection ID to address
 	connToAddr map[int64]pubcluster.Address
+	addrs      []pubcluster.Address
 }
 
-func newConnectionMap() *connectionMap {
+func newConnectionMap(lb pubcluster.LoadBalancer) *connectionMap {
 	return &connectionMap{
-		connectionsMu: &sync.RWMutex{},
-		addrToConn:    map[pubcluster.Address]*Connection{},
-		connToAddr:    map[int64]pubcluster.Address{},
+		lb:         lb,
+		mu:         &sync.RWMutex{},
+		addrToConn: map[pubcluster.Address]*Connection{},
+		connToAddr: map[int64]pubcluster.Address{},
 	}
 }
 
 func (m *connectionMap) AddConnection(conn *Connection, addr pubcluster.Address) {
-	m.connectionsMu.Lock()
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	// if the connection was already added, skip it
+	if _, ok := m.connToAddr[conn.connectionID]; ok {
+		return
+	}
 	m.addrToConn[addr] = conn
 	m.connToAddr[conn.connectionID] = addr
-	m.connectionsMu.Unlock()
+	m.addrs = append(m.addrs, addr)
 }
 
 // RemoveConnection removes a connection and returns true if there are no more addrToConn left.
 func (m *connectionMap) RemoveConnection(removedConn *Connection) int {
-	m.connectionsMu.Lock()
+	m.mu.Lock()
 	var remaining int
 	for addr, conn := range m.addrToConn {
 		if conn.connectionID == removedConn.connectionID {
 			delete(m.addrToConn, addr)
 			delete(m.connToAddr, conn.connectionID)
+			m.removeAddr(addr)
 			break
 		}
 	}
 	remaining = len(m.addrToConn)
-	m.connectionsMu.Unlock()
+	m.mu.Unlock()
 	return remaining
 }
 
 func (m *connectionMap) CloseAll() {
-	m.connectionsMu.RLock()
+	m.mu.RLock()
 	for _, conn := range m.addrToConn {
 		conn.close(nil)
 	}
-	m.connectionsMu.RUnlock()
+	m.mu.RUnlock()
 }
 
 func (m *connectionMap) GetConnectionForAddr(addr pubcluster.Address) *Connection {
-	m.connectionsMu.RLock()
+	m.mu.RLock()
 	conn := m.addrToConn[addr]
-	m.connectionsMu.RUnlock()
+	m.mu.RUnlock()
 	return conn
 }
 
 func (m *connectionMap) GetAddrForConnectionID(connID int64) (pubcluster.Address, bool) {
-	m.connectionsMu.RLock()
+	m.mu.RLock()
 	addr, ok := m.connToAddr[connID]
-	m.connectionsMu.RUnlock()
+	m.mu.RUnlock()
 	return addr, ok
 }
 
 func (m *connectionMap) RandomConn() *Connection {
-	m.connectionsMu.RLock()
-	var conn *Connection
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+	if len(m.addrs) == 0 {
+		return nil
+	}
+	addr := m.lb.OneOf(m.addrs)
+	conn := m.addrToConn[addr]
+	if conn != nil {
+		return conn
+	}
```

Comment:
Should we need to check here, if client is smart?

Reply:
When the client is non-smart, there is only a single connection in connection map, so `OneOf` should return the only address in the map. But I've made that more explicit at: https://github.com/hazelcast/hazelcast-go-client/pull/580/commits/8ff2f4a9dc5ac6228040a9b24a7045c40fa06bcd